### PR TITLE
Improve error message for partitioning by nested field in Iceberg

### LINKED ISSUE #15091: Improve error message when trying to partition on an Iceberg nested field
Partitioning on a field of a row type is not supported, but when attempting to do this you currently get a NullPointerException rather than a clear error message.

Slack thread: https://trinodb.slack.com/archives/CJ6UC075E/p1668717668592149

```
CREATE TABLE iceberg.my_db.my_table(a ROW(b VARCHAR)) WITH (format = 'PARQUET', partitioning = ARRAY['"a.b"'])
```

### PR DESCRIPTION
**Pull request by ebyhr**
## Description

Fixes #15091

### CODE REVIEW

**Comment by ebyhr** [comment]
CI hit #14441

**Review by homar — APPROVED** [review]

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java
Comment (findinpath) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java:977]: Just to be on the safe side, we can also add an extra assertion with a 2 level nesting - grandparent.parent.child - I tried it and works as expected.

**Review by alexjo2144 — APPROVED** [review]
If you don't mind adding some Spark compatibility test cases, here or in a separate PR:

- Create a table in Spark partitioned on a nested field
  - Trino should be able to read it
  - Trino should not allow insert/update/delete/merge operations on it

**Comment by ebyhr** [comment]
Fixing `TestIcebergSparkCompatibility.testPartitionedByNestedFiled ` now.

**Comment by ebyhr** [comment]
@alexjo2144 It seems the connector couldn't read such tables. I added another commit to fix this issue and updated more code to handle other statements. Could you take another look when you have time?

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java
Comment (findinpath) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:490]: I find the current solution a bit hard to understand just by reading the code. It is rather unclear why the `columns` is supposed to be empty when `partitionSourceIds` is not empty. 
Maybe adding an extra comment would help?

Unrelated to the current PR: When Trino is going to be able to deal with nested fields as partitions is the output of `getColumns(...)` method supposed to change?

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergUtil.java
Comment (findinpath) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergUtil.java:681]: I'd be rather inclined to duplicate this check within `io.trino.plugin.iceberg.IcebergMetadata#getWriteLayout` then to make this method `public`. 
The `indexParents` map parameter is rather abstract.

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergSparkCompatibility.java
Comment (findinpath) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergSparkCompatibility.java:449]: Out of curiosity: In a previous version of the PR before this last commit, the `DELETE` operation was still working. Why did it get removed?

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java
Comment (ebyhr) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:489]: > why the columns is supposed to be empty when partitionSourceIds is not empty.

Do you mean "columns is **_not_** supposed to be empty when `partitionSourceIds` is not empty"?

Yes, the method `getColumns` will be changed when we support nested fields as partitions.

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergSparkCompatibility.java
Comment (ebyhr) [reply testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergSparkCompatibility.java:449]: Relates to #15178. My branch status was before 84848cbafb85c4c78f1b48ebff2e4d7d289dccb1

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java
Comment (findinpath) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:489]: > Do you mean "columns is not supposed to be empty when partitionSourceIds is not empty"?

I mean that something that would hint a code maintainer that the partition source ids may point to a nested field which will not be returned as a column handle may by `getColumns(..)` may help.

**Review by findinpath — APPROVED** [review]

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java
Comment (alexjo2144) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:489]: I think the existing name `columns` is just not descriptive. Maybe something like `identityPartitionColumns`?

It also may be a redundant check. If `partitionSourceIds` is not empty it's impossible for `columns` to be empty. So the second block of the if here could be a `verify` instead.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSinkProvider.java
Comment (alexjo2144) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSinkProvider.java:87]: Sink setup feels a little late in the query execution to be catching this. Can we have these checks in IcebergMetadata, [like this one](https://github.com/trinodb/trino/blob/master/plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java#L1711)?

**Review by alexjo2144 — COMMENTED** [review]
Should we wait for a comment on if this should be supported from the Iceberg folks before merging the first commit? https://github.com/trinodb/trino/issues/15109#issuecomment-1327730105

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java
Comment (ebyhr) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:489]: >  If partitionSourceIds is not empty it's impossible for columns to be empty.

No, that's possible in case of nested fields.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSinkProvider.java
Comment (ebyhr) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSinkProvider.java]: Removed the 1st commit.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java
Comment (electrum) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:1827]: Nit: add a colon before the column name

```java
throw new TrinoException(NOT_SUPPORTED, "Partitioning by nested field is unsupported: " + field.name());

**Review by electrum — APPROVED** [review]

**Pull request merged**
